Redesign Build Deployment Process (External)

MukuFlash03 opened the pull request:
This PR serves as the implementation for [this issue](https://github.com/e-mission/e-mission-docs/issues/1048) for redesigning our build and deployment processes across e-mission-server, join, admin-dash, public-dash primarily.

Collaborating on this along with @nataliejschultz .

All four redesign PRs for the external repos:
E-mission-server: https://github.com/e-mission/e-mission-server/pull/961
Join: https://github.com/e-mission/nrel-openpath-join-page/pull/29
Admin-dash: https://github.com/e-mission/op-admin-dashboard/pull/113
Public-dash: https://github.com/e-mission/em-public-dashboard/pull/125

-------------

A rough plan can be:

1st phase
- Consolidating Differences
  - Minimize the differences in internal and external versions, cleanup unnecessary files, so on.
- Image_Build_Push.yml: 
  - CI / CD via GitHub actions to build and push images to Dockerhub for all four repos based on the current process for e-mission-server.
  - Image build push only builds for that specific repo when merge is made to a specific branch.
  - Additionally, for e-mission-server, image-build-push will now also need to trigger the GitHub action in join, admin-dash, public-dash.

2nd phase: 
- Multi-tier uniform repo structure in nrelopenpath
  - Need to change the internal repos to pull from the externally built images for each repo without duplicating the entire external code repository internally.

MukuFlash03 commented:
Checking in the first initial set of changes for consolidating the differences in the external and internal versions.
These are based on the [differences identified here](https://github.com/e-mission/e-mission-docs/issues/1048#issuecomment-1998477929)

Current changes involve:

1. Sed changed to jq
Changed it in this script viz_scripts/docker/start_noteboook.sh to match the way it is being done internally.
TODO: Might not even need to do this, once we decide how to handle environment variables, and might get rid of using the respective conf files at all.

-----------

These changes are still pending:
viz_scripts/docker/start_noteboook.sh: Python notebooks run directly internally instead of cron tab
viz_scripts/Dockerfile: AWS certificates.

nataliejschultz commented on viz_scripts/Dockerfile:
```diff
@@ -1,5 +1,7 @@
-FROM shankari/e-mission-server:master_2024-02-10--19-38
+# Please change once all PR changes are final, so it reads from shankari/e-mission-server
+# FROM shankari/e-mission-server:master_2024-02-10--19-38
+FROM mukuflash03/e-mission-server:image-push-merge_2024-04-16--49-36
```
Reminder that we will want to use a different account for these images in the future

nataliejschultz commented on viz_scripts/docker/Dockerfile.dev:
```diff
@@ -1,5 +1,7 @@
 # python 3
-FROM shankari/e-mission-server:master_2024-02-10--19-38
+# FROM shankari/e-mission-server:master_2024-02-10--19-38
+# Please change once all PR changes are final, so it reads from shankari/e-mission-server
```
Reminder to change before merging

nataliejschultz commented on viz_scripts/docker/cert.sh:
```diff
@@ -0,0 +1,6 @@
+if [ -z ${PROD_STAGE} ] ; then
+    echo “Not in staging / production environment, continuing build...” 
```
This script allows us to eliminate a major difference between internal and external repos by allowing an environment variable to control whether or not we download the certificate.

shankari replied:
I see. However, I don't think this check is necessary. The certificate bundle is public so it is not sensitive information. Having the certs in stage/production doesn't harm anything, they are just an extraneous file which will be sitting around doing nothing.

And removing the check and always copying the file will make this much simpler - we don't need to have a separate environment variable that is set sometimes and not at other times.

nataliejschultz replied:
I can delete cert.sh and add a line to the dockerfile of viz-scripts to copy the file, just like how it is in the internal repo. I'll also remove the code where the script is run from.

nataliejschultz reviewed (commented):
More reminders and comment for why we need the script.

MukuFlash03 commented:
Docker Volume usage

When testing initially, I was unable to see generated plots in the frontend dashboard page.
This is when I learnt about how volumes defined in docker-compose actually work.

So, I was testing with the internal repos,
- Created folders in internal nrelopenpath multi-tier.
    - Created public_dash directory and moved frontend and viz_scripts inside it.
    - Created .gitignored common plots directory in public-dash.
   
- Added -v flags to both frontend and viz_scripts containers docker run commands.
- This creates two-way connection - any changes in either of the directories before and after colon in -v flag will be updated when changes occur in the other.
    - Thus, viz_scripts should update and store plots in its internal docker plots/
    - Due to volume connection, this should update and also store plots in external local or host plots/
    - Again, frontend is also connected to the external volume, so it should also have access to plots/ in its internal docker public/plots/

----------

Great! It’s working. Plots visible.
- Plots do have the label - “Unable to generate plots”, though which come up as empty tables with 0 (zero) values.
    - Hopefully it’s there for internal repo main branch as well.
        - Yes, same message of missing plots. Perhaps unable to read from db as no data loaded. [Confirmed, this was the reason.]

- Run “docker run” command inside nrelopenpath/public_dash/ directory, else plots/ created outside public_dash in root nrelopenpath folder.
    - No issues as such, but let’s keep all public-dash stuff inside its own directory.

MukuFlash03 commented:
### Testing Public-dash the right way by loading data 

@iantei  [documentation](https://github.com/e-mission/e-mission-docs/blob/master/docs/dev/publicdashboard/Working_with_public_dashboard.md) was really helpful:

I have been testing the external repo with the image-push branch, which contains the redesign code changes.
The Dockerfile uses the base server image from my personal Dockerhub repository.

Earlier, while I was spinning up the DB container as well, I wasn't actually loading any data in it.
Without data, I was seeing empty data tables with 0 values, with only year-month values being present.
Additionally, a text in red was present in the plot - "Unable to generate plot ; reason:"

-----

Steps followed to test code changes:

A. Built images and ran containers using docker-compose:
```
Dev
$ docker compose -f docker-compose.dev.yml build 
$ docker compose -f docker-compose.dev.yml up -d

Prod
$ docker compose -f docker-compose.yml build 
$ docker compose -f docker-compose.yml up -d
```


B. This time, I loaded stage data using available stage snapshot dataset once the containers were up:
```
$ bash viz_scripts/docker/load_mongodump.sh /Users/mmahadik/Documents/Data/openpath-stage-snapshot-dec-20.tar.gz
```

C. Executing Jupyter notebooks

Before executing the notebook, I had to change the year and month in the notebooks to None instead of current hardcoded values year = 2020 and month = 11.

Then executed Jupyter notebooks following steps in this public-dashboard execution [documentation](https://github.com/e-mission/e-mission-docs/blob/master/docs/dev/publicdashboard/Working_with_public_dashboard.md#execution-of-different-public-dashboard-notebooks-automatically-without-launching-jupyter-notebook).

---------

MukuFlash03 commented:
**Testing observations for production docker compose file** 
I was successfully able to load data and see plots with actual data like staging public-dash [URL](https://openpath-stage.nrel.gov/public/)

Screenshots showing generated plots

Public-dash External

<img width="1792" alt="public-dash-ext-1" src="https://github.com/e-mission/em-public-dashboard/assets/37911144/e869eef7-1220-43c1-b46a-0b735d250056">


Public-dash Staging [for reference]

<img width="1792" alt="public-dash_staging" src="https://github.com/e-mission/em-public-dashboard/assets/37911144/d9cb946b-8244-4ad9-8c47-0efa66222286">

nataliejschultz commented:
I have spent many hours trying to test em-public-dashboard (from current main branch as a positive control) with the openpath_prod_uue_snapshot_oct3.tar.gz.

I followed the procedure that was outlined by the documentation, as well as by @MukuFlash03 above, and still only got the empty result:

<img width="1010" alt="Screenshot 2024-04-24 at 11 02 59 AM" src="https://github.com/e-mission/em-public-dashboard/assets/90212258/565bd994-1b05-43e0-9d68-3fcd57a43dd3">

Mukul tested the procedure using this same dataset, and got the exact same result. I am going to seek out another dataset to test with.

nataliejschultz commented:
Testing of the public dash has been really difficult. I have tried so, so many times, and have finally gotten the current published version of public dash to generate images.

On my local copy of the public dash published code, I did the following:

Set all of the year and month values in each relevant ipynb to None (despite the [documentation](https://github.com/e-mission/e-mission-docs/blob/master/docs/dev/publicdashboard/Working_with_public_dashboard.md#execution-of-different-public-dashboard-notebooks-automatically-without-launching-jupyter-notebook) saying not to do this?)

Changing DB_HOST=mongodb://db/openpath_stage in the docker compose + dev
Using STUDY_CONFIG=stage-program (I am using data that corresponds to this)

Running: 

```
$ docker compose -f docker-compose.dev.yml build
$ docker compose -f docker-compose.dev.yml up -d 
$ bash viz_scripts/docker/load_mongodump.sh ../../openpath-stage-snapshot-***-**.tar.gz
$docker exec -it em-public-dashboard-notebook-server-1 /bin/bash
```
Inside viz-scripts container:

```
source setup/activate.sh
cd saved-notebooks/
PYTHONPATH=.. python bin/update_mappings.py mapping_dictionaries.ipynb
PYTHONPATH=.. python bin/generate_plots.py generic_metrics.ipynb default
```
Opening the localhost at port 3274: 

<img width="980" alt="Screenshot 2024-04-27 at 2 45 42 AM" src="https://github.com/e-mission/em-public-dashboard/assets/90212258/4aa00406-b16a-42b0-a637-46756ece7d44">



On my local copy of Mukul’s image-push branch, I wanted to now try the cortezebikes data. Here is what I did:

Set all of the year and month values in each relevant ipynb to None 

Create db container:
$ docker run --name db -d -p 27017:27017 --network emission mongo:4.4.0

Frontend:
```
$ docker build -t int-public-dash-frontend ./frontend/

$ docker run --name int-public-dash-frontend-1 -d -p 3274:6060 -v ./plots:/public/plots --network emission int-public-dash-frontend
```
Viz-scripts:
Changed to build from my local image of our changed server code:
<img width="648" alt="Screenshot 2024-04-27 at 3 03 09 AM" src="https://github.com/e-mission/em-public-dashboard/assets/90212258/c5a7ef23-15b7-45e7-be63-c906ce8d6219">


Ran:
```
$ docker build -t viz-scripts ./viz_scripts/

$ docker run --name int-public-dash-notebook-1 -d -e DB_HOST=[mongodb://db/](mongodb://db/openpath_stage)openpath_stage -e WEB_SERVER_HOST=0.0.0.0 -e STUDY_CONFIG=cortezebikes -p 47962:8888 -v ./plots:/plots --network emission viz-scripts
$ docker exec -it int-public-dash-notebook-1 /bin/bash
```
Inside viz-scripts container:

```
source setup/activate.sh
cd saved-notebooks
PYTHONPATH=.. python bin/update_mappings.py mapping_dictionaries.ipynb
PYTHONPATH=.. python bin/generate_plots.py generic_metrics.ipynb default
```
<img width="615" alt="Screenshot 2024-04-27 at 4 16 59 AM" src="https://github.com/e-mission/em-public-dashboard/assets/90212258/b74c56a6-3cc3-4b0e-9e68-b516f1c740b2">


This loaded some data! 

Of course, I also wanted to try to recreate the stage data result from before.  

I followed the exact same steps as above, using a different STUDY_CONFIG and DB_HOST:

```
$ docker build -t viz-scripts ./viz_scripts/
$ docker run --name int-public-dash-notebook-1 -d -e DB_HOST=mongodb://db/openpath_stage -e WEB_SERVER_HOST=0.0.0.0 -e STUDY_CONFIG=stage-program -p 47962:8888 -v ./plots:/plots --network emission viz-scripts
```
<img width="860" alt="Screenshot 2024-04-27 at 3 39 50 AM" src="https://github.com/e-mission/em-public-dashboard/assets/90212258/2b9cacef-a459-4130-b699-b7e906afd987">

It did not work. I am unsure as to why, and am at a bit of a loss for what to do from here.

shankari commented:
> On my local copy of Mukul’s image-push branch, I wanted to now try some program-specific data. Here is what I did:
...
> `$ docker run --name int-public-dash-notebook-1 -d -e DB_HOST=[mongodb://db/](mongodb://db/openpath_stage)openpath_stage -e WEB_SERVER_HOST=0.0.0.0 -e STUDY_CONFIG=cortezebikes -p 47962:8888 -v ./plots:/plots --network emission viz-scripts`

Why are you not using `docker-compose` as before? If you change both the image and the process to run, any error that you encounter might be the result of either change.

> This loaded some data!

Or it may just be showing results from the previous successful run. Did you clear out the `plots` directory between runs?

shankari commented:
> However, when I tried this with the dev compose file, got this error in edb get_database.py file where it says that config["url"] raises a KeyError.

This is almost certainly because the directory that you are running from does not have a `conf/storage/db.conf` file. But I thought we were not going to use the `conf/storage/db.conf` any more and were going to use environment variables instead. Has that change been implemented?

nataliejschultz commented:
> Why are you not using `docker-compose` as before? If you change both the image and the process to run, any error that you encounter might be the result of either change.
 
In reference to [this comment](https://github.com/e-mission/e-mission-server/pull/961#issuecomment-2076405167), I wanted to make sure that I was building everything off of locally generated images. I will try using this locally built image with the docker compose instead of docker run. However, in the internal repos, we have to test with docker run since the Jenkins pipeline cannot use compose. 


> > This loaded some data!
> 
> Or it may just be showing results from the previous successful run. Did you clear out the `plots` directory between runs?

No, I thought these plots would be overwritten. Comparing the "Number of trips All data Default" graph for the stage data (left), the cortezebikes snapshot that I loaded into my locally built container (middle), and the current public [cortezebikes public dashboard ](https://cortezebikes-openpath.nrel.gov/public/)(right), you can see that my locally generated image in the middle is consistent with the cortezebikes trip data up to the recent snapshot date.

<img width="936" alt="Screenshot 2024-04-29 at 12 13 25 PM" src="https://github.com/e-mission/em-public-dashboard/assets/90212258/b04ebda3-4b42-4012-870b-643d822196ee">

> What did you try to debug? Did you try to look at logs? Did you try to add print statements? Did you try to look at the documentation for the public dashboard?

In debugging, I have tried: 

1. Referring to the [public dashboard readme](https://github.com/e-mission/em-public-dashboard/blob/main/README.md) and [developer documentation](https://github.com/e-mission/e-mission-docs/blob/master/docs/dev/publicdashboard/Working_with_public_dashboard.md), as well as the process outlined by @iantei 
2. Using `docker exec -it` to change various variables to read the data in properly, ie: `export STUDY_CONFIG=stage-program`, `export DB_HOST=DB_HOST=mongodb://db/openpath_stage` and using `echo $STUDY_CONFIG` to make sure exporting the variables took before re-running the notebooks
3. Downloading vim into the viz-scripts container while using docker exec (`apt-get install vim`) and directly modifying the ipynb files via vi + insert+ wq, using cat command to make sure changes took, and re-running the notebook run commands listed above
4. Viewing the mongo and notebook container logs in the docker desktop 
5. Using `docker exec -it` inside the mongo container, and running `mongo` then
`show dbs` to make sure the data has loaded and to check that the db name matches with my DB_HOST environment variable

MukuFlash03 commented:
> > However, when I tried this with the dev compose file, got this error in edb get_database.py file where it says that config["url"] raises a KeyError.
> 
> This is almost certainly because the directory that you are running from does not have a `conf/storage/db.conf` file. But I thought we were not going to use the `conf/storage/db.conf` any more and were going to use environment variables instead. Has that change been implemented?

Yes, the changes have been implemented in e-mission-server so that environment variables are used instead of the config files. This issue is not being observed in admin-dash, or public-dash when loading using production docker compose file.

The issue was just seen with the dev version of the docker-compose.dev file and that too on running the plots for the python notebooks manually. 

When I directly connected to the container via the terminal and executed commands which invoked edb, the "url" value was printed correctly.

I've mentioned my debugging process in this [comment above](https://github.com/e-mission/em-public-dashboard/pull/125#issuecomment-2075298292).

shankari commented:
@MukuFlash03 I did see your debugging process. Other that investigating the code myself, my only hint is that this is almost certainly because the file is missing. Your question should then be: why is the dev version using the file instead of the environment variable?

nataliejschultz commented:
Re-testing my local version of public dashboard using docker compose this time. Pre-run checks:

1. Dockerfile is set to build my local server image with changes: consolidate-diff-server-localbuild: 
<img width="378" alt="Screenshot 2024-04-29 at 5 04 49 PM" src="https://github.com/e-mission/em-public-dashboard/assets/90212258/fffa747b-b7a0-4ff9-bfaf-2b8df6d2a1b4">


2. Docker-compose.yml has DB_HOST=mongodb://db/openpath_stage and STUDY_CONFIG=stage-program:
<img width="402" alt="Screenshot 2024-04-29 at 3 37 32 PM" src="https://github.com/e-mission/em-public-dashboard/assets/90212258/f7537427-4783-462d-abf1-2b5acea45b93">

3. All ipynb notebooks have year and month set to none: <img width="1744" alt="Screenshot 2024-04-29 at 3 34 48 PM" src="https://github.com/e-mission/em-public-dashboard/assets/90212258/02dbe560-4cb6-49b5-8395-9b7c1cd7cd15">

4. Plots folder emptied
<img width="176" alt="Screenshot 2024-04-29 at 3 37 52 PM" src="https://github.com/e-mission/em-public-dashboard/assets/90212258/7a55ee2f-fac6-4191-9020-eea0d649b273">

5. Other containers utilizing the same ports removed

Commands run: 
```
$ docker compose -f docker-compose.yml build
$ docker compose -f docker-compose.yml up -d
$ docker ps
$ docker rename mukul-forked-em-public-dashboard-db-1 em-public-dashboard-db-1
$ bash viz_scripts/docker/load_mongodump.sh ../../openpath-stage-snapshot-dec-20.tar.gz 
$ docker exec -it mukul-forked-em-public-dashboard-notebook-server-1 /bin/bash
root@6a1431d35053:/usr/src/app# source setup/activate.sh 
(emission) root@6a1431d35053:/usr/src/app# cd saved-notebooks/
(emission) root@6a1431d35053:/usr/src/app/saved-notebooks# PYTHONPATH=.. python bin/update_mappings.py mapping_dictionaries.ipynb
(emission) root@6a1431d35053:/usr/src/app/saved-notebooks# PYTHONPATH=.. python bin/generate_plots.py generic_metrics.ipynb default
```
<img width="1155" alt="Screenshot 2024-04-29 at 4 45 02 PM" src="https://github.com/e-mission/em-public-dashboard/assets/90212258/40e810a0-2576-4216-a6d0-202fc1515e4e">

**Docker compose with the locally built server image from the consolidate-differences branch worked!**

Reflecting on my previous attempts , this is what's different between my previous test using `docker build`:

1. My build command didn't include CRON_MODE=TRUE. Would have to investigate more to see if this is the root of the problem.
2. I was not re-building the db container between every run. The container is automatically re-built in the compose, so I had to start fresh this time. 
3. I was not clearing out the plots volume/folder.

Since I will have to use docker build to test the internal build, I will likely find out if (1) was the issue all along.

MukuFlash03 commented:
> Your question should then be: why is the dev version using the file instead of the environment variable?

Found the reason for the bug and fixed it in this commit in the server repo.
https://github.com/MukuFlash03/e-mission-server/commit/e778b3f3cf050eec33ef439a1b6763b6aaf533e0

The bug was triggered by a `.gitignore-d conf/ directory` in public-dash which had a `db.conf file`.
This was being loaded when `docker-compose.dev.yml` was used to test the dev version of public-dash.

This was occurring since there was a nested dictionary in the db.conf.sample and db.conf files while I had initially only stored the nested keys (url, result_limit). Since it was still reading from the file, it stored the nested dictionary format with timeseries as the parent key followed by (url, result_limit) as children.

Hence, the correct way to access the db.conf values would be `["timeseries"]["url"]` instead of just `["url"]`.
Fixed it by accessing these values correctly.

nataliejschultz commented:
Testing the Internal repo. Pre-run checks:

1. Dockerfile is set to build my locally built images from previous run in which I used docker compose
2. Mongo container running

These images should have the values in ipynb notebooks set to none, since those changes are contained in the images. 

Frontend: 
```
$ docker build -t int-public-dash-frontend ./public-dashboard/frontend/
$ docker run --name int-public-dash-frontend-1 -d -p 3274:6060 -v ./plots:/public/plots --network emission em-pub-dash-prod/frontend
```
Checked to make sure frontend built properly:
<img width="978" alt="Screenshot 2024-04-30 at 2 37 17 PM" src="https://github.com/e-mission/em-public-dashboard/assets/90212258/33e9576f-0990-410e-ad64-77db7d6cc39e">

Viz-scripts:
```
$  docker build -t int-public-dash-notebook ./public-dashboard/viz_scripts/
$ docker run --name int-public-dash-notebook-1 -d -e DB_HOST=mongodb://db/openpath-stage -e WEB_SERVER_HOST=0.0.0.0 -e CRON_MODE=TRUE -e STUDY_CONFIG="stage-program" -p 47962:8888 -v ./plots:/plots --network emission em-pub-dash-prod/viz-scripts
$ docker exec -it int-public-dash-notebook-1 /bin/bash
root@3d727ce04e7d:/usr/src/app# source setup/activate.sh 
(emission) root@3d727ce04e7d:/usr/src/app# cd saved-notebooks/
(emission) root@3d727ce04e7d:/usr/src/app/saved-notebooks# PYTHONPATH=.. python bin/update_mappings.py mapping_dictionaries.ipynb
(emission) root@3d727ce04e7d:/usr/src/app/saved-notebooks# PYTHONPATH=.. python bin/generate_plots.py generic_metrics.ipynb default
```
<img width="860" alt="Screenshot 2024-04-27 at 3 39 50 AM" src="https://github.com/e-mission/em-public-dashboard/assets/90212258/0404dc97-1c92-482c-8e30-1eae5bd12369">

It did not load the images, just the default “unable to generate plot” images. I’ve spent a long time trying to figure out what could have gone wrong. 

The root of the issue appears to be that the ipynb can’t extract data from the db. Noted differences between the internal and external testing processes:

1. Cron mode is not set to anything internally, and a large portion of start_script.sh is commented out and replaced with code to run AWS scheduled tasks
2. The db is not generated automatically, as it is with docker compose. It seems like there is a difference between how the db name is parsed, since the em-public-dash-db-1 is found automatically with the external build but the internal requires a db name of DB. To try to debug this, I’ve added prints to some of the scripts in the container with vim. Ex:
```
CURRENT DATABASE URL: mongodb://db/openpath-stage
CURRENT_DB:  Database(MongoClient(host=['db:27017'], document_class=dict, tz_aware=False, connect=True, uuidrepresentation=3), 'openpath-stage')
```
Which has not been very helpful yet. I also tried exporting the DB_HOST variable differently and STUDY_CONFIG, and have not had luck yet. I've followed the process of getting the db name using get_database.py, and how it interacts with the scaffolding.py script to try to understand what could be happening along the way.

I want to discuss with Mukul about how his process went, and if he has successfully gotten the images to generate. If that isn't an option I will simply keep pushing and trying different inputs/prints/container modifications

 Potential causes of this issue: path, parsing of name, cron schedule being removed for AWS

Today I’m going to try: 
- Using the other data set to see what happens, and if this doesn't elucidate the problem:
- Re-building the server image with the changes from [this commit ](https://github.com/e-mission/e-mission-server/pull/961/commits/a0190d431a1c4b11fa1abb6b26793f9ab027859a). I have already used vim to incorporate these changes into my container, but I will start from scratch if I have to.

nataliejschultz commented:
Testing the Internal repo one more time, the same way as [here.](https://github.com/e-mission/em-public-dashboard/pull/125#issuecomment-2088985965)

Again, I deleted the db container and re-load the data:

```
$ docker run --name db -d -p 27017:27017 --network emission mongo:4.4.0
$ bash public-dashboard/viz_scripts/docker/load_mongodump.sh ../../openpath-stage-snapshot-dec-20.tar.gz
```
And reset the environment variables (but note, am using the same container):
<img width="714" alt="Screenshot 2024-05-01 at 1 45 23 PM" src="https://github.com/e-mission/em-public-dashboard/assets/90212258/755983dc-15a1-4b86-8916-e6789ffdcfe2">

I cleared out the plots folder, and re-ran the notebooks. 

<img width="997" alt="Screenshot 2024-05-01 at 2 05 31 PM" src="https://github.com/e-mission/em-public-dashboard/assets/90212258/5f6a5aa4-72e6-4f7b-b96f-847b7ad075d3">

I had to empty the safari cache to load the right images, but it finally worked!!!!


It seems like the db container has an issue with finding the correct data when there are two data dumps present. Re-creating the container every time fixed it.

shankari commented on viz_scripts/docker/cert.sh:
where is this script launched from?

shankari commented:
@nataliejschultz please also resolve the merge conflicts

nataliejschultz commented:
> @nataliejschultz please also resolve the merge conflicts. Note also that this is still set to draft.

Resolved conflicts for merge

nataliejschultz commented:
I removed cert.sh and modified the viz_scripts dockerfile to copy it every time, rather than based on the environment variable. All other tags/directories/etc are prepared for merge with main

shankari commented on .github/workflows/image_build_push.yml:
```diff
@@ -0,0 +1,144 @@
+
+      - name: Install Python dependencies
+        run: |
+         python -m pip install --upgrade pip
```
nit: Is `pip` not installed on Github Action images?
All the ubuntu installations have pip
https://github.com/actions/runner-images/blob/main/images/ubuntu

shankari commented on .github/workflows/image_build_push.yml:
```diff
@@ -0,0 +1,144 @@
+        docker_image_tag: ${{ steps.get_docker_tag.outputs.docker_image_tag }}
+
+      steps:
+      - uses: actions/checkout@v4
```
nit, future fix: why are we checking this out twice?

nataliejschultz replied:
Based on[ this](https://github.com/actions/checkout/issues/19#issuecomment-523866061), each job will need its own checkout/environment

shankari commented on .github/workflows/image_build_push.yml:
```diff
@@ -0,0 +1,144 @@
+    runs-on: ubuntu-latest
+
+    env:
+      DOCKER_IMAGE_TAG_1: ${{ needs.fetch_tag.outputs.docker_image_tag }}
+      DOCKER_IMAGE_TAG_2: ${{ github.event.inputs.docker_image_tag }}
```
nit, future fix: better names

nataliejschultz replied:
Clarification: Better names for the environment variables?

shankari commented on .github/workflows/image_build_push.yml:
```diff
@@ -0,0 +1,144 @@
+
+    - name: Add, Commit, Push changes to .env file
+      run: |
+        git config --local user.email "action@github.com"
+        git config --local user.name "GitHub Action"
```
nit, future fix: configure this to be more meaningful; "GitHub Action" doing what?

shankari commented on .github/workflows/image_build_push.yml:
```diff
@@ -0,0 +1,144 @@
+    - name: build docker image
+      run: |
+        if [ "${{ github.event_name }}" == "workflow_dispatch" ]; then
+          docker build --build-arg DOCKER_IMAGE_TAG=$DOCKER_IMAGE_TAG_2 -t $DOCKER_USER/${GITHUB_REPOSITORY#*/}:${GITHUB_REF##*/}_${{ steps.date.outputs.date }} ./frontend
+          docker build --build-arg DOCKER_IMAGE_TAG=$DOCKER_IMAGE_TAG_2 -t $DOCKER_USER/${GITHUB_REPOSITORY#*/}_notebook:${GITHUB_REF##*/}_${{ steps.date.outputs.date }} ./viz_scripts
+        else
+          docker build --build-arg DOCKER_IMAGE_TAG=$DOCKER_IMAGE_TAG_1 -t $DOCKER_USER/${GITHUB_REPOSITORY#*/}:${GITHUB_REF##*/}_${{ steps.date.outputs.date }} ./frontend
+          docker build --build-arg DOCKER_IMAGE_TAG=$DOCKER_IMAGE_TAG_1 -t $DOCKER_USER/${GITHUB_REPOSITORY#*/}_notebook:${GITHUB_REF##*/}_${{ steps.date.outputs.date }} ./viz_scripts
```
we should not be building the images using `docker build`
The build using GitHub actions and the README should be consistent.
The GitHub action should validate that the README and the README instructions are correct.
We should follow the instructions in the README and build using `docker-compose -f docker-compose.yml build`

nataliejschultz replied:
Okay, I changed this to use docker compose. In order to do so, I used an environment variable to set the names of the images being built with compose to the proper image tag name, so that the subsequent image push command will work. The desired tag name was previously passed into the build command, which is not possible with docker compose. 

I also renamed the variable `DOCKER_IMAGE_TAG` to `SERVER_IMAGE_TAG`, as I think this is a more meaningful variable name and is helpful for discerning all of the different tags.

shankari commented on docker-compose.dev.yml:
```diff
@@ -3,8 +3,8 @@ services:
     build:
-        context: frontend
-        dockerfile: docker/Dockerfile.dev
+      context: frontend
+      dockerfile: docker/Dockerfile.dev
```
whitespace changes

shankari commented on docker-compose.yml:
```diff
@@ -2,7 +2,9 @@ version: "3"
   dashboard:
     image: em-pub-dash-prod/frontend
-    build: frontend
+    build: 
+      context: frontend
+      dockerfile: Dockerfile
```
I don't think we need this change. The default dockerfile is `Dockerfile`

shankari commented on viz_scripts/docker/start_notebook.sh:
```diff
@@ -5,9 +5,8 @@
-    sed "s-localhost-${local_host}_" conf/storage/db.conf.sample > conf/storage/db.conf
-else
-    sed "s-localhost-${DB_HOST}-" conf/storage/db.conf.sample > conf/storage/db.conf
+    export DB_HOST=$local_host
+    echo "Setting db host environment variable to localhost"
```
nit, future fix: this can be removed and simplified (similar to https://github.com/e-mission/e-mission-server/pull/961#discussion_r1605946966)

shankari reviewed (changes requested):
I am fine with merging this for now under the principle of working to working.
But for the next round of polishing, I would like to understand why we need to fetch the runID or do any of this complicated behavior given that we pass in the image tag when this workflow is launched 
https://github.com/e-mission/e-mission-server/pull/961/files#diff-bde90ebb933051b12f18fdcfcefe9ed31e2e3950d416ac84aec628f1f9cc2780R77

Please also note: https://github.com/e-mission/e-mission-docs/issues/1048#issuecomment-2119333182

shankari reviewed (changes requested):
@MukuFlash03 @nataliejschultz you need to resolve the merge conflicts. There are also several minor issues that should be polished. If you can get to them by EOD tomorrow (end of Tue, 18 Aug) that would be good, so that I can merge this and move on to the next set of changes to review.

nataliejschultz commented on .github/workflows/image_build_push.yml:
```diff
@@ -0,0 +1,158 @@
+        docker push $DOCKER_USER/${GITHUB_REPOSITORY#*/}:${GITHUB_REF##*/}_${{ steps.date.outputs.date }}
+        docker push $DOCKER_USER/${GITHUB_REPOSITORY#*/}_notebook:${GITHUB_REF##*/}_${{ steps.date.outputs.date }}
+   
+    - name: Create a text file
+      run: |
+          echo ${{ steps.date.outputs.date }} > public_dash_tag_file.txt
+          echo "Created tag text file"
+
+    - name: Upload Artifact
+      uses: actions/upload-artifact@v4
+      with:
+          name: public-dash-image-tag
+          path: public_dash_tag_file.txt
+          overwrite: true
```
@shankari the point of the artifact is to allow the internal repo to access the new tag outside of the run (see the[ internal repo PR](https://github.nrel.gov/nrel-cloud-computing/nrelopenpath/pull/50))

shankari replied:
I see. I still think we should be able to use tags and `.env` files instead, but will defer that discussion until the cleanup/polish

MukuFlash03 commented on viz_scripts/docker/start_notebook.sh:
This file does use this variable in setting the PYTHONPATH.
```
if [ -z ${CRON_MODE} ] ; then
    echo "Running notebook in docker, change host:port to localhost:47962 in the URL below"
    PYTHONPATH=/usr/src/app jupyter notebook --no-browser --ip=${WEB_SERVER_HOST} --allow-root
else
```

shankari replied:
`WEB_SERVER_HOST` is always set to `0.0.0.0`. we can just hardcode it
https://github.com/search?q=repo%3Ae-mission%2Fem-public-dashboard%20WEB_SERVER_HOST&type=code
https://github.com/e-mission/e-mission-server/pull/961#issuecomment-2272509467
https://github.com/e-mission/e-mission-server/commit/f65f94c8064b7a29033e64620ff7fa6f60bd99f8

MukuFlash03 commented on .github/workflows/image_build_push.yml:
```diff
@@ -0,0 +1,158 @@
+      - name: Use Run ID from previous fetch_run_id job
+        run: echo Run ID from previous job ${{ env.RUN_ID }}
+        
+      - name: Download artifact
+        uses: actions/download-artifact@v4
+        with:
+          # TODO: Create a token with basic repo permissions
+          name: docker-image-tag
+          github-token: ${{ secrets.GH_PAT_TAG }}
+          repository: e-mission/e-mission-server
+          run-id: ${{ env.RUN_ID }}
```
As highlighted in this [comment in the documented issue](https://github.com/e-mission/e-mission-docs/issues/1048#issuecomment-2093776437), one of the final stages I had reached in designing the CI/CD was to eliminate having to manually update any docker tags (from a developer's perspective who will use the repos).

In the public-dash and admin-dash workflows, workflow dispatch directly receives it via input parameters, so it doesn't have to deal with artifacts.
But the push event is the one that utilizes the docker tags from the downloaded artifact. 
The tag from this artifact is the one that is written to the .env file for every push event.


------

Now a few scenarios:
1. What happens if a developer doesn't have .env file in their checkout out repo?
The .env file would always be present once this PR has been merged.
As a part of this PR, we are manually adding the .env file with the latest server image tag currently available in Dockerhub. This manual addition only needs to be done once during these code changes. Once implemented, the workflow will automatically update the .env file. 
Developers would then just pull in the remote changes and they should have the appropriate .env file.

2. What happens if a developer changes the image tag in .env file?
The workflow first executes the steps to fetch the latest tag from the server repo.
Any existing tag in .env file will be overridden with this latest tag.
This ensures two things: 
a. incorrect/non-existent tags will be overridden.
b. latest server image is used always.

------

To summarize, no one would need to update the .env file, the workflow would do it automatically.

MukuFlash03 replied:
With that said, I do kind of see why we don't need to download the artifact.

Let's consider a few more scenarios:

1. Workflow dispatch: Server image is updated and new version pushed.
This will trigger the workflow dispatch which triggers workflows in the dashboard repos.
This updates the .env file.
The new tag is then used in building the image.

-----

2. Push event: Server image is not updated; dashboard repos have changes pushed.
In the existing workflow, the push will download the artifact and overwrite the .env file.

But, the .env file already has the latest tag from when the workflow dispatch event occurred. 
So, the push event can simply read in the latest server tag from the .env file.

MukuFlash03 commented on .github/workflows/image_build_push.yml:
```diff
@@ -0,0 +1,158 @@
+          echo "Workflow_dispatch: New server image built and pushed, Updating image tag in .env"
+          echo "SERVER_IMAGE_TAG=$DOCKER_IMAGE_TAG_2" > .env
+        else
+          echo "Push event: Restoring latest server image tag in .env"
+          echo "SERVER_IMAGE_TAG=$DOCKER_IMAGE_TAG_1" > .env
```
Right, understood ; there's no need to change .env file.
Workflow dispatch would have already updated the .env file whenever a new server version is released.
Detailed [comments here](https://github.com/e-mission/em-public-dashboard/pull/125#discussion_r1716150918).

MukuFlash03 commented on docker-compose.dev.yml:
```diff
@@ -18,8 +18,10 @@ services:
-        dockerfile: docker/Dockerfile.dev
+      context: viz_scripts
+      dockerfile: docker/Dockerfile.dev
+      args:
+        SERVER_IMAGE_TAG: ${SERVER_IMAGE_TAG}
```
No, the user will not be required to manually set the docker tags at any point. 
Unless they want to test other server image versions locally.

The args will be read in from the .env file.
I tested this locally by ensuring the .env file was present and had an image tag pointing to the correct server image.

It builds the image successfully.

-----

.env file with the latest server image tag
```
SERVER_IMAGE_TAG=2024-08-12--15-15
```

Some logs from docker compose command:
```
[+] Building 169.1s (21/21) FINISHED                                                              docker:desktop-linux
 => [notebook-server internal] load build definition from Dockerfile.dev                                          0.0s
 => => transferring dockerfile: 717B                                                                              0.0s
 => [dashboard internal] load build definition from Dockerfile.dev                                                0.0s
 => => transferring dockerfile: 178B                                                                              0.0s
 => [notebook-server internal] load metadata for docker.io/shankari/e-mission-server:master_2024-08-12--15-15     1.5s
 => [dashboard internal] load metadata for docker.io/library/node:21.7.2-alpine                                   1.6s
 => [notebook-server internal] load .dockerignore                                                                 0.0s
 => => transferring context: 2B                                                                                   0.0s
 => [dashboard internal] load .dockerignore                                                                       0.0s
 => => transferring context: 2B                                                                                   0.0s
 => [notebook-server 1/8] FROM docker.io/shankari/e-mission-server:master_2024-08-12--15-15@sha256:f05c128a8fe4  66.6s
```

MukuFlash03 commented on .github/workflows/image_build_push.yml:
```diff
@@ -0,0 +1,158 @@
+
+    - name: rename docker images
+      run: |
+        docker image tag em-pub-dash-prod/frontend:latest $DOCKER_USER/${GITHUB_REPOSITORY#*/}:${GITHUB_REF##*/}_${{ steps.date.outputs.date }}
```
Right, we shouldn't be doing this. 
I've added conditional checks for the `rename` and `push` steps in [this commit](https://github.com/MukuFlash03/em-public-dashboard/commit/d2a5f69784227be24a4cf38d95b4fc84881f1c74).

shankari commented on .github/workflows/image_build_push.yml:
```diff
@@ -129,20 +70,24 @@ jobs:
-        docker image tag em-pub-dash-prod/frontend:latest $DOCKER_USER/${GITHUB_REPOSITORY#*/}:${GITHUB_REF##*/}_${{ steps.date.outputs.date }}
-        docker image tag em-pub-dash-prod/viz-scripts:latest $DOCKER_USER/${GITHUB_REPOSITORY#*/}_notebook:${GITHUB_REF##*/}_${{ steps.date.outputs.date }}
+        if [ "${{ github.event_name }}" == "push" ]; then
+          docker image tag em-pub-dash/frontend:latest $DOCKER_USER/${GITHUB_REPOSITORY#*/}:${GITHUB_REF##*/}_${{ steps.date.outputs.date }}
```
Let's make sure that both the notebook image and the webserver image are labeled. Otherwise just seeing a "dashboard" image will cause confusion.

```suggestion
          docker image tag em-pub-dash/frontend:latest $DOCKER_USER/${GITHUB_REPOSITORY#*/}_frontend:${GITHUB_REF##*/}_${{ steps.date.outputs.date }}
```

shankari commented on .github/workflows/image_build_push.yml:
```diff
@@ -129,20 +70,24 @@ jobs:
       run: |
-        docker push $DOCKER_USER/${GITHUB_REPOSITORY#*/}:${GITHUB_REF##*/}_${{ steps.date.outputs.date }}
+        if [ "${{ github.event_name }}" == "push" ]; then
+          docker push $DOCKER_USER/${GITHUB_REPOSITORY#*/}:${GITHUB_REF##*/}_${{ steps.date.outputs.date }}
```
```suggestion
          docker push $DOCKER_USER/${GITHUB_REPOSITORY#*/}_frontend:${GITHUB_REF##*/}_${{ steps.date.outputs.date }}
```

shankari reviewed (changes requested):
@MukuFlash03 one more minor change. But changing the repo name later will be complicated, so let's try to get this right the first time.

shankari commented:
Squash-merging again (similar to https://github.com/e-mission/e-mission-server/pull/961#issuecomment-2284918011) since otherwise we have 73 commits for 8 files.

shankari commented:
Test fails: https://github.com/e-mission/em-public-dashboard/actions/runs/10393542152

MukuFlash03 commented:
@nataliejschultz had added [the commit](https://github.com/e-mission/em-public-dashboard/pull/125/commits/82cf4e9463f81c0dfff1e72328a1661875b5ea38) but it was using the incorrect file name. 

So I changed the file name in [this commit](https://github.com/e-mission/em-public-dashboard/pull/125/commits/7483f2adff2e162d4ca454e0a275af2b7cbfbe31). But I ended up changing the image name as well in [this commit](https://github.com/e-mission/em-public-dashboard/pull/125/commits/f325eb35b406824ac4db1155546706fa45d333ec).

shankari commented:
@MukuFlash03 Note that, with these changes, the frontend and notebook images will have different tags. While you are changing this, you may want to change the final step to upload an artifact with two tags (or two artifacts, one for each image). @nataliejschultz will then have to change the internal script to handle both tags

MukuFlash03 commented:
> @MukuFlash03 Note that, with these changes, the frontend and notebook images will have different tags. While you are changing this, you may want to change the final step to upload an artifact with two tags (or two artifacts, one for each image). @nataliejschultz will then have to change the internal script to handle both tags

We do not need changes for this since the `tag` is the timestamp generated:

```
    - name: push docker images
      run: |
        if [ "${{ github.event_name }}" == "push" ]; then
          docker push $DOCKER_USER/${GITHUB_REPOSITORY#*/}_frontend:${GITHUB_REF##*/}_${{ steps.date.outputs.date }}
        fi
        docker push $DOCKER_USER/${GITHUB_REPOSITORY#*/}_notebook:${GITHUB_REF##*/}_${{ steps.date.outputs.date }}
   
    - name: Create a text file
      run: |
          echo ${{ steps.date.outputs.date }} > public_dash_tag_file.txt
          echo "Created tag text file"
```

This is the same for all images in the workflow run. So internally, the image name without the timestamp differs for frontend and notebook images which is hardcoded anyways as it is static. Only the timestamp tag changes based on the workflow run.

shankari commented:
> This is the same for all images in the workflow run. So internally, the image name without the timestamp differs for frontend and notebook images which is hardcoded anyways as it is static. Only the timestamp tag changes based on the workflow run.

For the `workflow_dispatch` runs, it is not.
Consider the following:
- push: both images are tagged with `2024-08-14`
- new server build on 2024-08-16, `viz_script` is tagged with `2024-08-16`, frontend is still at  `2024-08-14`

MukuFlash03 commented:
I see. 
In that case, @nataliejschultz can handle this in the internal script by checking for the type of trigger event.
I did inspect the response from the GitHub REST API ([URL ](https://api.github.com/repos/MukuFlash03/em-public-dashboard/actions/workflows/90180351/runs)for my branch workflows)

The event type is available:
```
      "run_number": 49,
      "event": "push",
      "status": "completed",
      "conclusion": "success",
      "workflow_id": 90180351,
....
....
....
      "run_number": 31,
      "event": "workflow_dispatch",
      "status": "completed",
      "conclusion": "success",
      "workflow_id": 90180351,
``

shankari commented:
@MukuFlash03 Just because something is possible, doesn't mean that it is the elegant approach.

If you see the script that @nataliejschultz I think it would be a lot easier to just download two archive files. Note that she has a common function to download the archive file which is invoked once for each tag. having to looking at the type for only this repo would make that script a lot more ugly.

MukuFlash03 commented:
I tried implementing the logic for handling two different tags for frontend and notebook but it's not correct.
I'm trying to think it through but we don't maintain the separate tags for frontend / notebook in the workflow currently.

-----

For push events, we can have two artifacts. They'll have the same timestamps as tags.
For workflow, for the notebook image we'll use the latest timestamp for the 1st artifact.
 
But what should be passed as the 2nd artifact for frontend image? 
We'll need to retrieve the last stored tag for frontend from somewhere?

MukuFlash03 commented:
I tested the end-to-end flow with workflow dispatch from the server repo in these runs:
- [server](https://github.com/MukuFlash03/e-mission-server/actions/runs/10395538299), [public-dash](https://github.com/MukuFlash03/em-public-dashboard/actions/runs/10395578575), [admin-dash](https://github.com/MukuFlash03/op-admin-dashboard/actions/runs/10395578638).

Well, two tags are generated in the public-dash workflow run, but the tags are wrong. 

Notebook tag is correct since it is fetched from current timestamp in either case.
 
But frontend tag is storing the server image tag.
It should store last tagged frontend image.

-------

So, as mentioned in this comment, I'm thinking how to retrieve the last tag for frontend image. This will need to be stored / maintained somewhere as well.

shankari commented:
you can just save it in a different `.env` file which is overridden on push but not on workflow_dispatch
Or you can remove the optimization and rebuild both images every time, arguing that it makes the downstream flow easier and that the frontend image size is small (you have to check and actually make that argument)

MukuFlash03 commented:
Frontend image is small ~ 50 MB [Source: My dockerhub account]
But I've gone ahead and implemented it using a different .env file.

<img width="1792" alt="Screenshot 2024-08-14 at 3 43 43 PM" src="https://github.com/user-attachments/assets/5c435b7b-0882-48a9-af4c-d187123a1fc4">

MukuFlash03 commented:
1. Push event [run](https://github.com/MukuFlash03/em-public-dashboard/actions/runs/10396428921)

This workflow run was triggered by a `PUSH` event.
The tags were updated correctly and two artifacts produced (FRONTEND, NOTEBOOK).

------

2. Workflow Dispatch run [[server](https://github.com/MukuFlash03/e-mission-server/actions/runs/10396495809), [public-dash](https://github.com/MukuFlash03/em-public-dashboard/actions/runs/10396543869), [admin-dash](https://github.com/MukuFlash03/op-admin-dashboard/actions/runs/10396543873)]

Next, I triggered a workflow run from e-mission-server via `Workflow Dispatch` event.

This was able to read in the current FRONTEND image tag from the `.env.repoTags` file and upload the existing value as an artifact.
NOTEBOOK tag on the other hand is updated normally as it should irrespective of event type.
